### PR TITLE
Squares API exercise

### PR DESCRIPTION
**Pull request by indremike**
A few comments:
1. Persistence layer was created as a dummy, just to avoid using external database services and save time for this exercise. Alternatively, I would have used MongoDb (or other type of database) Docker container for local development.
2. As a next step I would have written unit tests.
3. The algorithm for detecting the squares I think is not optimal but I was not really focusing at the performance of it in this exercise.
4. Not all the required API endpoints were implemented but it would not be very difficult to implement the remaining ones because the most interfaces are already created.
5. My priority was to build all the abstractions at first, so it would be easier to implement any new functionality.
6. It took me about 8 hours to build this service from scratch.

### CODE REVIEW

File: Squares.Api/Controllers/SquaresController.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
+using Microsoft.AspNetCore.Mvc;
+using Squares.Api.Presenters;
+using Squares.Application.Dtos;
+using Squares.Application.Interfaces.UseCases;
+using Squares.Application.Requests;
+using Squares.Application.Responses;
+using Squares.Application.UseCases;
+using System.ComponentModel.DataAnnotations;
+using System.Drawing;
```
Comment (R3purg) [inline Squares.Api/Controllers/SquaresController.cs:9]: Unused `using` statements should be deleted for improved code cleanliness:
```
using Squares.Application.Responses;
using Squares.Application.UseCases;
using System.Drawing;
```

File: Squares.Api/Controllers/SquaresController.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            SquaresDetectionEntriesListPresenter identifiedSquaresListingPresenter)
+        {
+            _logger = logger;
+            _createNewPointsListUseCase = createNewPointsListUseCase;
+            _getIdentifiedSquaresListingUseCase = getIdentifiedSquaresListingUseCase;
+            _addNewPointToDetectionSquaresEntryUseCase = addNewPointToDetectionSquaresEntryUseCase;
+            _removePointFromDetectionSquaresEntryUseCase = removePointFromDetectionSquaresEntryUseCase;
+            _identifiedSquaresPresenter = identifiedSquaresPresenter;
+            _identifiedSquaresListingPresenter = identifiedSquaresListingPresenter;
+        }
```
Comment (R3purg) [inline Squares.Api/Controllers/SquaresController.cs:40]: Optional (or Consider): Separate blocks of code with an empty `new line`.

File: Squares.Api/Controllers/SquaresController.cs
Comment (R3purg) [inline Squares.Api/Controllers/SquaresController.cs:1]: FYI: The controller class could extend some `ApiBaseController`, that implements all the needed attributes and RESTful conventions, so every time a new endpoint is implemented, it would not be necessary to duplicate same attributes.

Example of `ApiBaseController`:
```
[ApiController]
[Route("api/v1/[controller]/")]
public class ApiBaseController : ControllerBase
{
	// Base functions to use in controllers
}
```

This would make the application more scalable for future implementations.

File: Squares.Api/Controllers/SquaresController.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+using System.ComponentModel.DataAnnotations;
+using System.Drawing;
+
+namespace Squares.Api.Controllers
+{
+    [ApiController]
+    [Route("[controller]")]
+    public class SquaresController : ControllerBase
+    {
+        private readonly ILogger<SquaresController> _logger;
```
Comment (R3purg) [inline Squares.Api/Controllers/SquaresController.cs:17]: Optional (or Consider): `_logger` is not used as of right now in this controller, consider integrating it more or exclude it entirely.

File: Squares.Api/Dockerfile
Comment (R3purg) [inline Squares.Api/Dockerfile:1]: Dockerfile is incomplete, missing `COPY` of other projects:

- Squares.Application
- Squares.Domain
- Squares.Infrastructure

File: Squares.Api/MappingProfiles/DomainToDtoMappingProfile.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+﻿using AutoMapper;
+using Squares.Application.Dtos;
+using Squares.Domain.Entities;
+using System;
```
Comment (R3purg) [inline Squares.Api/MappingProfiles/DomainToDtoMappingProfile.cs:4]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Api/Presenters/SquaresDetectionEntriesListPresenter.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+﻿using Squares.Application.Dtos;
+using Squares.Application.Interfaces;
+using Squares.Application.Responses;
+using System.Net;
+
+namespace Squares.Api.Presenters
+{
+    public class SquaresDetectionEntriesListPresenter : IOutputPort<SquaresDetectionEntriesListResponse>
+    {
+        public IList<SquaresDetectionEntryDto> SquaresDetectionEntriesList { get; private set; }
```
Comment (R3purg) [inline Squares.Api/Presenters/SquaresDetectionEntriesListPresenter.cs:10]: It would be a good idea if the `SquaresDetectionEntriesList` property was initialized to an empty list to avoid null reference exceptions.

File: Squares.Api/Presenters/SquaresDetectionEntryPresenter.cs
Comment (R3purg) [inline Squares.Api/Presenters/SquaresDetectionEntryPresenter.cs:1]: FYI: If the `SquaresDetectionEntryPresenter` and `SquaresDetectionEntriesListPresenter` will always share the same properties and behavior, they could be combined into a single class, implementing some base class, that handles the functionality.

Make sure to follow SOLID principles.

File: Squares.Api/Program.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
+using AutoMapper;
```
Comment (R3purg) [inline Squares.Api/Program.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using AutoMapper;
using Microsoft.Extensions.DependencyInjection;
using Swashbuckle.Swagger;
```

File: Squares.Application/Dtos/PointDto.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Application/Dtos/PointDto.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Dtos/SquareDto.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Application/Dtos/SquareDto.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Dtos/PointDto.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Squares.Application.Dtos
+{
+    public struct PointDto
+    {
+        public int X { get; set; }
```
Comment (R3purg) [inline Squares.Application/Dtos/PointDto.cs:11]: Nit: Since this is a `struct`, making `PointDto` immutable will prevent the state of the `struct` to change after it is created. This could improve code robustness.

Optional (or Consider): Since `PointDto` is used as a reference type, it is not necessary to have it as a `struct`, unless there are scalability ideas ahead. Otherwise, it may be better to use a class instaed.

File: Squares.Application/Dtos/SquaresDetectionEntryDto.cs [hunk 10]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Application/Dtos/SquaresDetectionEntryDto.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Dtos/SquaresDetectionEntryDto.cs [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Squares.Application.Dtos
+{
+    public class SquaresDetectionEntryDto
+    {
+        public string Id { get; set; }
+        public IList<PointDto> InputPoints { get; set; }
```
Comment (R3purg) [inline Squares.Application/Dtos/SquaresDetectionEntryDto.cs:12]: It would be a good idea if the `InputPoints` & `DetectedSquares` properties were initialized to an empty list to avoid null reference exceptions.

File: Squares.Application/Interfaces/IOutputPort.cs [hunk 12]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Application/Interfaces/IOutputPort.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Interfaces/Repositories/IRepository.cs [hunk 13]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Application/Interfaces/Repositories/IRepository.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Interfaces/Repositories/ISquaresDetectionEntriesRepository.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
+﻿using Squares.Domain.Entities;
+using System;
```
Comment (R3purg) [inline Squares.Application/Interfaces/Repositories/ISquaresDetectionEntriesRepository.cs:2]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Interfaces/UseCases/IAddNewPointToDetectionSquaresEntryUseCase.cs [hunk 15]
```diff
@@ -0,0 +1,14 @@
+﻿using Squares.Application.Requests;
+using Squares.Application.Responses;
+using System;
```
Comment (R3purg) [inline Squares.Application/Interfaces/UseCases/IAddNewPointToDetectionSquaresEntryUseCase.cs:3]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Interfaces/UseCases/ICreateSquaresDetectionEntryUseCase.cs [hunk 16]
```diff
@@ -0,0 +1,14 @@
+﻿using Squares.Application.Requests;
+using Squares.Application.Responses;
+using System;
```
Comment (R3purg) [inline Squares.Application/Interfaces/UseCases/ICreateSquaresDetectionEntryUseCase.cs:3]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Interfaces/UseCases/IGetSquaresDetectionEntriesListUseCase.cs [hunk 17]
```diff
@@ -0,0 +1,14 @@
+﻿using Squares.Application.Requests;
+using Squares.Application.Responses;
+using System;
```
Comment (R3purg) [inline Squares.Application/Interfaces/UseCases/IGetSquaresDetectionEntriesListUseCase.cs:3]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Interfaces/UseCases/ICreateSquaresDetectionEntryUseCase.cs
Comment (R3purg) [inline Squares.Application/Interfaces/UseCases/ICreateSquaresDetectionEntryUseCase.cs:1]: FYI: If the `SquaresDetectionEntryPresenter` and `SquaresDetectionEntriesListPresenter` will always share the same properties and behavior, interfaces `ICreateSquaresDetectionEntryUseCase` & `IGetSquaresDetectionEntriesListUseCase` could be combined.

File: Squares.Application/Interfaces/UseCases/IRemovePointFromDetectionSquaresEntryUseCase.cs [hunk 18]
```diff
@@ -0,0 +1,14 @@
+﻿using Squares.Application.Requests;
+using Squares.Application.Responses;
+using System;
```
Comment (R3purg) [inline Squares.Application/Interfaces/UseCases/IRemovePointFromDetectionSquaresEntryUseCase.cs:3]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Interfaces/UseCases/IUseCaseRequestHandler.cs [hunk 19]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Application/Interfaces/UseCases/IUseCaseRequestHandler.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Requests/CreateSquaresDetectionEntryRequest.cs [hunk 20]
```diff
@@ -0,0 +1,19 @@
+﻿using Squares.Application.Dtos;
+using Squares.Application.Responses;
+using System;
```
Comment (R3purg) [inline Squares.Application/Requests/CreateSquaresDetectionEntryRequest.cs:3]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Requests/CreateSquaresDetectionEntryRequest.cs [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Squares.Application.Requests
+{
+    public class CreateSquaresDetectionEntryRequest : IUseCaseRequest<SquaresDetectionEntryResponse>
+    {
+        public List<PointDto> Points { get;}
```
Comment (R3purg) [inline Squares.Application/Requests/CreateSquaresDetectionEntryRequest.cs:13]: It would be a good idea if the `Points` property was initialized to an empty list to avoid null reference exceptions.

For example it is called without a check in `CreateNewPointsListUseCase.Handle(...)`.

Consider initializing it like so:
```
public IList<PointDto> Points { get; } = new List<PointDto>();

public CreateSquaresDetectionEntryRequest(List<PointDto> points) 
{
	Points = points;
}
```

File: Squares.Application/Requests/GetISquaresDetectionEntriesListRequest.cs [hunk 22]
```diff
@@ -0,0 +1,18 @@
+﻿using Squares.Application.Dtos;
+using Squares.Application.Responses;
+using System;
```
Comment (R3purg) [inline Squares.Application/Requests/GetISquaresDetectionEntriesListRequest.cs:3]: Unused `using` statements should be deleted for improved code cleanliness:
```
using Squares.Application.Dtos;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Requests/IUseCaseRequest.cs [hunk 23]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Application/Requests/IUseCaseRequest.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Requests/RemoveOrAddPointInDetectionSquaresEntryRequest.cs [hunk 24]
```diff
@@ -0,0 +1,21 @@
+﻿using Squares.Application.Dtos;
+using Squares.Application.Responses;
+using System;
```
Comment (R3purg) [inline Squares.Application/Requests/RemoveOrAddPointInDetectionSquaresEntryRequest.cs:3]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Requests/RemoveOrAddPointInDetectionSquaresEntryRequest.cs [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Squares.Application.Requests
+{
+    public class RemoveOrAddPointInDetectionSquaresEntryRequest : IUseCaseRequest<SquaresDetectionEntryResponse>
+    {
+        public string EntryId { get; }
```
Comment (R3purg) [inline Squares.Application/Requests/RemoveOrAddPointInDetectionSquaresEntryRequest.cs:13]: Optional (or Consider): `Id` would be a good name, instead of `EntryId`.

`RemoveOrAddPointInDetectionSquaresEntryRequest` class name is quite long, might cause confusion. Consider renaming it into something more concise.

File: Squares.Application/Responses/SquaresDetectionEntriesListResponse.cs [hunk 26]
```diff
@@ -0,0 +1,22 @@
+﻿using Squares.Application.Dtos;
+using System;
```
Comment (R3purg) [inline Squares.Application/Responses/SquaresDetectionEntriesListResponse.cs:2]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Responses/SquaresDetectionEntriesListResponse.cs [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Squares.Application.Responses
+{
+    public class SquaresDetectionEntriesListResponse : UseCaseResponseMessage
+    {
+        public IList<SquaresDetectionEntryDto> SquaresDetectionEntriesList { get; }
```
Comment (R3purg) [inline Squares.Application/Responses/SquaresDetectionEntriesListResponse.cs:13]: It would be a good idea if the `SquaresDetectionEntriesList` property was initialized to an empty list to avoid null reference exceptions.

File: Squares.Application/Responses/SquaresDetectionEntryResponse.cs [hunk 28]
```diff
@@ -0,0 +1,22 @@
+﻿using Squares.Application.Dtos;
+using System;
```
Comment (R3purg) [inline Squares.Application/Responses/SquaresDetectionEntryResponse.cs:2]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/Responses/SquaresDetectionEntriesListResponse.cs
Comment (R3purg) [inline Squares.Application/Responses/SquaresDetectionEntriesListResponse.cs:1]: FYI: If the `SquaresDetectionEntriesListResponse` and `SquaresDetectionEntryResponse` will always share the same properties and behavior, they could be combined into a single class, implementing some base class, that handles the functionality.

File: Squares.Application/Responses/UseCaseResponseMessage.cs [hunk 29]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Application/Responses/UseCaseResponseMessage.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/UseCases/AddNewPointToDetectionSquaresEntryUseCase.cs [hunk 30]
```diff
@@ -0,0 +1,53 @@
+﻿using AutoMapper;
+using Squares.Application.Dtos;
+using Squares.Application.Interfaces;
+using Squares.Application.Interfaces.Repositories;
+using Squares.Application.Interfaces.UseCases;
+using Squares.Application.Requests;
+using Squares.Application.Responses;
+using Squares.Domain.Entities;
+using System;
```
Comment (R3purg) [inline Squares.Application/UseCases/AddNewPointToDetectionSquaresEntryUseCase.cs:9]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/UseCases/AddNewPointToDetectionSquaresEntryUseCase.cs [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+        private readonly IMapper _mapper;
+        private readonly ISquaresDetectionEntriesRepository _identifiedSquaresRepository;
+
+        public AddNewPointToDetectionSquaresEntryUseCase(IMapper mapper, ISquaresDetectionEntriesRepository identifiedSquaresRepository)
+        {
+            _mapper = mapper;
+            _identifiedSquaresRepository = identifiedSquaresRepository;
+        }
+
+        public bool Handle(RemoveOrAddPointInDetectionSquaresEntryRequest request, IOutputPort<SquaresDetectionEntryResponse> outputPort)
```
Comment (R3purg) [inline Squares.Application/UseCases/AddNewPointToDetectionSquaresEntryUseCase.cs:28]: FYI (If in the future the application will use a database): Method `Handle` should have a return type of `Task<bool>` instead of `bool`, and the method should be made `async`.

File: Squares.Application/UseCases/AddNewPointToDetectionSquaresEntryUseCase.cs [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+        {
+            var entry = _identifiedSquaresRepository.GetByIdAsync(request.EntryId);
```
Comment (R3purg) [inline Squares.Application/UseCases/AddNewPointToDetectionSquaresEntryUseCase.cs:30]: FYI (If in the future the application will use a database): `GetByIdAsync` is named as an asynchronous functions, however it is not the case in the `SquaresDetectionEntriesRepository`, this should be changed to have an asynchronous operation.

File: Squares.Application/UseCases/AddNewPointToDetectionSquaresEntryUseCase.cs [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+
+            if (entry == null) {
+                outputPort.Handle(new SquaresDetectionEntryResponse(System.Net.HttpStatusCode.NotFound, $"DetectionSquaresEntry with id {request.EntryId} was not found"));
+                return false;
+            }
+            if (entry.InputPoints.Contains(_mapper.Map<Point>(request.Point)))
```
Comment (R3purg) [inline Squares.Application/UseCases/AddNewPointToDetectionSquaresEntryUseCase.cs:36]: Here a list of `InputPoints` is called. This can cause a null reference exception. This is why it is important to initialize an empty list (**check previous comments**).

File: Squares.Application/UseCases/CreateNewPointsListUseCase.cs [hunk 34]
```diff
@@ -0,0 +1,40 @@
+﻿using AutoMapper;
+using Squares.Application.Dtos;
+using Squares.Application.Interfaces;
+using Squares.Application.Interfaces.Repositories;
+using Squares.Application.Interfaces.UseCases;
+using Squares.Application.Requests;
+using Squares.Application.Responses;
+using Squares.Domain.Entities;
+using System;
```
Comment (R3purg) [inline Squares.Application/UseCases/CreateNewPointsListUseCase.cs:9]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/UseCases/CreateNewPointsListUseCase.cs [hunk 35]
```diff
@@ -0,0 +1,40 @@
...
+        private readonly IMapper _mapper;
+        private readonly ISquaresDetectionEntriesRepository _identifiedSquaresRepository;
+        public CreateNewPointsListUseCase(IMapper mapper, ISquaresDetectionEntriesRepository identifiedSquaresRepository)
+        {
+            _mapper = mapper;
+            _identifiedSquaresRepository = identifiedSquaresRepository;
+        }
+        public bool Handle(CreateSquaresDetectionEntryRequest request, IOutputPort<SquaresDetectionEntryResponse> outputPort)
+        {
+            var newSqaure = new SquaresDetectionEntry(Guid.NewGuid().ToString(), _mapper.Map<List<PointDto>, List<Point>>(request.Points.Distinct().ToList()));
```
Comment (R3purg) [inline Squares.Application/UseCases/CreateNewPointsListUseCase.cs:30]: Check if the `request.Points` parameter is not null before converting it to avoid a null reference exception.

File: Squares.Application/UseCases/GetIdentifiedSquaresListingUseCase.cs [hunk 36]
```diff
@@ -0,0 +1,42 @@
+﻿using AutoMapper;
+using Squares.Application.Dtos;
+using Squares.Application.Interfaces;
+using Squares.Application.Interfaces.Repositories;
+using Squares.Application.Interfaces.UseCases;
+using Squares.Application.Requests;
+using Squares.Application.Responses;
+using System;
```
Comment (R3purg) [inline Squares.Application/UseCases/GetIdentifiedSquaresListingUseCase.cs:8]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Application/UseCases/GetIdentifiedSquaresListingUseCase.cs [hunk 37]
```diff
@@ -0,0 +1,42 @@
...
+        public GetIdentifiedSquaresListingUseCase(ISquaresDetectionEntriesRepository identifiedSquaresRepository, IMapper mapper)
+        {
+            _identifiedSquaresRepository = identifiedSquaresRepository;
+            _mapper = mapper;
+        }
+        public bool Handle(GetISquaresDetectionEntriesListRequest request, IOutputPort<SquaresDetectionEntriesListResponse> outputPort)
+        {
+            try
+            {
+                outputPort.Handle(new SquaresDetectionEntriesListResponse(_mapper.Map<IList<SquaresDetectionEntryDto>>(_identifiedSquaresRepository.GetAllAsync())));
```
Comment (R3purg) [inline Squares.Application/UseCases/GetIdentifiedSquaresListingUseCase.cs:31]: FYI (If in the future the application will use a database): `GetAllAsync` should be awaited, since it is named as an asynchronous operation. It should also be changed to an asynchronous operation in the repository class or renamed accordingly.

File: Squares.Application/Requests/GetISquaresDetectionEntriesListRequest.cs [hunk 38]
```diff
@@ -0,0 +1,18 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Squares.Application.Requests
+{
+    public class GetISquaresDetectionEntriesListRequest : IUseCaseRequest<SquaresDetectionEntriesListResponse>
```
Comment (R3purg) [inline Squares.Application/Requests/GetISquaresDetectionEntriesListRequest.cs:11]: Confusing class name `GetISquaresDetectionEntriesListRequest`, should be changed to `GetSquaresDetectionEntriesListRequest` or something even more concise to avoid confusion.

File: Squares.Application/UseCases/RemovePointFromDetectionSquaresEntryUseCase.cs [hunk 39]
```diff
@@ -0,0 +1,31 @@
+﻿using AutoMapper;
+using Squares.Application.Interfaces;
+using Squares.Application.Interfaces.Repositories;
+using Squares.Application.Interfaces.UseCases;
+using Squares.Application.Requests;
+using Squares.Application.Responses;
+using System;
```
Comment (R3purg) [inline Squares.Application/UseCases/RemovePointFromDetectionSquaresEntryUseCase.cs:7]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Domain/Entities/Point.cs [hunk 40]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Domain/Entities/Point.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Domain/Entities/Square.cs [hunk 41]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Domain/Entities/Square.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Domain/Entities/Square.cs [hunk 42]
```diff
@@ -0,0 +1,24 @@
...
+    public class Square
+    {
+        public Point PointA { get; set; }
+        public Point PointB { get; set; }
+        public Point PointC { get; set; }
+        public Point PointD { get; set; }
+
+        public Square(Point pointA, Point pointB, Point pointC, Point pointD)
+        {
+            this.PointA = pointA;
```
Comment (R3purg) [inline Squares.Domain/Entities/Square.cs:18]: If the `Point` class instances can change after being assigned to the `Square` instance, this could lead to unexpected behavior. It might be considerable to make the `Point` class instances immutable.

Check `PointDto` code block comments above.

File: Squares.Domain/Entities/SquaresDetectionEntry.cs [hunk 43]
```diff
@@ -0,0 +1,73 @@
+﻿using System;
```
Comment (R3purg) [inline Squares.Domain/Entities/SquaresDetectionEntry.cs:1]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Domain/Entities/SquaresDetectionEntry.cs [hunk 44]
```diff
@@ -0,0 +1,73 @@
...
+                Dist(p3, p4)
+            };
+
+            return distinct.Count == 2 && !distinct.Contains(0);
+        }
+
+        /// <summary>
+        /// Distance between two points
+        /// </summary>
+        public static int Dist(Point p1, Point p2) => (p1.X - p2.X) * (p1.X - p2.X) + (p1.Y - p2.Y) * (p1.Y - p2.Y);
```
Comment (R3purg) [inline Squares.Domain/Entities/SquaresDetectionEntry.cs:71]: Nit: `Dist` function could be made as an extension to `Point` struct.

File: Squares.Domain/Entities/SquaresDetectionEntry.cs [hunk 45]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            get { return DetectedSquares.Count; }
+        }
+
+        public SquaresDetectionEntry(string id, IList<Point> inputPoints)
+        {
+            this.Id = id;
+            this.InputPoints = inputPoints;  
+            this.DetectedSquares = new List<Square>();  
+        }
```
Comment (R3purg) [inline Squares.Domain/Entities/SquaresDetectionEntry.cs:26]: Optional (or Consider): Separate blocks of code with an empty `new line`.

File: Squares.Domain/Entities/SquaresDetectionEntry.cs [hunk 46]
```diff
@@ -0,0 +1,73 @@
...
+        /// <summary>
+        /// Using brute force it checks if given four points make a square
+        /// </summary>
+        public void IdentifySquares()
```
Comment (R3purg) [inline Squares.Domain/Entities/SquaresDetectionEntry.cs:30]: Brute-force algorithm's time complexity (**big O**) can be simplified with `HashTable` type to store the points and their coordinates.

File: Squares.Domain/Entities/SquaresDetectionEntry.cs [hunk 47]
```diff
@@ -0,0 +1,73 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Squares.Domain.Entities
+{
+    public class SquaresDetectionEntry
+    {
+        public string Id { get; set; }
+        public IList<Point> InputPoints { get; set; }
```
Comment (R3purg) [inline Squares.Domain/Entities/SquaresDetectionEntry.cs:12]: It would be a good idea if the `InputPoints` & `DetectedSquares` properties were initialized to an empty list to avoid null reference exceptions.

Then in `line 25` it is not necessary anymore to initialize `DetectedSquares` property.

File: Squares.Infrastructure/Repositories/SquaresDetectionEntriesRepository.cs [hunk 48]
```diff
@@ -0,0 +1,51 @@
+﻿using Squares.Application.Interfaces.Repositories;
+using Squares.Domain.Entities;
+using System;
+using System.Collections.Concurrent;
```
Comment (R3purg) [inline Squares.Infrastructure/Repositories/SquaresDetectionEntriesRepository.cs:4]: Unused `using` statements should be deleted for improved code cleanliness:
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
```

File: Squares.Infrastructure/Repositories/SquaresDetectionEntriesRepository.cs
Comment (R3purg) [inline Squares.Infrastructure/Repositories/SquaresDetectionEntriesRepository.cs:1]: Exception handling is necessary to prevent data loss.

FYI: For scalability, if in the future the application should access a database, instead of `ConcurrentDictionary`, await/async principle would be necessary.

File: Squares.Infrastructure/Repositories/SquaresDetectionEntriesRepository.cs [hunk 49]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            IdentifiedSquares.TryAdd(entity.Id, entity);
+        }
+
+        public void DeleteAsync(string id)
+        {
+            IdentifiedSquares.TryRemove(id, out SquaresDetectionEntry entity);
+        }
+
+        public IList<SquaresDetectionEntry> GetAllAsync()
```
Comment (R3purg) [inline Squares.Infrastructure/Repositories/SquaresDetectionEntriesRepository.cs:34]: Optional (or Consider): Might be a good idea to consider pagination, since the retrieval of data could be too big to handle smoothly. This would make he application more scalable.

**Review by R3purg — CHANGES_REQUESTED** [review]
Since the repository is using a `ConcurrentDictionary`, it is thread-safe. However, in the future, if we want to scale the application further and implement it with a database layer, the async/await operations would be a better idea.
All of the related issues were marked with "**FYI (If in the future the application will use a database)**".

Class and method names should also be checked to make them more concise. Some of them are too long and confusing.

Exception handling is lacking in the main repository class `SquaresDetectionEntriesRepository`. Adding `try/catch` blocks would help.

Tests are not implemented at all, the application might be prone to errors and edge cases. Moreover, it is vulnerable to other attacks, this should be handled with **Security Headers**, check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers
